### DISCUSSION TITLE
Support for Adafruit I2C QT Rotary Encoder

### DISCUSSION THREAD

**Discussion by JMBMdes** [arendst/Tasmota#20271] [Ideas]
Hi all! Let me first thank you for your awesome work with Tasmota.

I'm triying to give a second live to an old bottle cooler that is not working because a broken thermostat. My idea is to control compresor and heater based in DS18B20 readings and show temperature on a TM1637 4 digit display. I have a running prototype using rules that I plan to move to berry.

I would like to change the target temperature using a rotary (yes, I could "hardcode" the target temperature, but that's no fun, right?  :-) 
To reuse the holes in the front panel, I'm thinking on a rotary, and what I've found is the [Adafruit I2C rotary encoder](https://learn.adafruit.com/adafruit-i2c-qt-rotary-encoder/overview). It uses the seesaw library, and I've seen it's implemented in Tasmota for soil sensor.
 
My question is if it is feasible to implement the rotary.

Thank you for your comments.

**Comment by sfromis** [comment]
Until someone decides to create a driver (which should be quite feasible), that I2C rotary encoder is not supported by Tasmota. 

What **is** supported is the simple pulse-based rotary encoder often called KY-040.

**Reply by barbudor** [reply]
If you select a board based on ESP32, including the lowest cost ESP32C3, you will be able to use the Berry language to write an I2C driver.
Berry is very simiar in syntax to (micro)python although not exactly the same, but anyone with a little bit of experience with Arduino or micropython should go through without much difficulties.

If you stick on the old, but still working, ESP8266, then you have the choice of the Scripting-Language for which the learning curve can be a little more longer

And in anycase, a native C/C++ driver will always work for any of the ESP models.

**Reply by JMBMdes** [reply]
Thanks for your answers.

Yep, I'm using an ESP32, and I've been practicing with Berry from a few weeks ago. I read the cookbook recipe to develop an I2C driver, but I think my current knowledge is not enough to address that "development". Hopefully someone finds it's worthy to include in Tasmota. It really works very well with the examples provided by adafruit: press buttom, rotary, neopixel as feefback...

> And in anycase, a native C/C++ driver will always work for any of the ESP models

You mean not using Tasmota but development from sscratch, right? Well then the logic and management of different features I need for the cooler will be more complicated for me...

**Reply by sfromis** [reply]
What he meant was creating a native C++ driver for Tasmota, to exist along with lots of other device drivers sharing the framework of Tasmota, and thereby available for setting up an useful combination. It is common for those drivers to not be written from scratch, but be based on or reusing an existing library.

Creating a driver with Berry would be from scratch, with the Berry language being more friendly to quick development, without the "minefield" C++ can be to developers not well-versed in the language.

**Reply by JMBMdes** [reply]
Got it, that was what I though (and why I wrote the support request...).  I'm afraid that, more than often, I dont't express myself very well in english...  :-(
Thanks.
Jose.

**Reply by arendst** [reply]
Why going over the trouble of writing a rotary encoder driver for the Adafruit I2C rotary encoder which doesn't even have a rotary encoder anyway.

Just get a rotary encoder en use the current Tasmota rotary encoder driver. 

Have fun.

**Reply by JMBMdes** [reply]
You made me laugh! I'm so embarrased... I just realize that my I2C QT Rotary Encoder is in reallity "an I2C encoder" for a rotary encoder that is actually missing...

Hope not all my participations here will be so "brilliant", thank you!

**Reply by sfromis** [reply]
Still, one benefit of using the "Adafruit I2C QT Rotary Encoder" could be supporting up to 8 of these on I2C, while the existing Tasmota rotary encoder driver only supports 2.

Maybe it could make sense for "someone" to create a Tasmota driver for a more general Adafruit Seesaw support. IDK how feasible or practical that could be, with a "I2C to something" converter.
